emulator/periph/src/flash_ctrl.rs

```diff
@@ -0,0 +1,1096 @@
+        &mut self,
+        _size: emulator_types::RvSize,
+        _val: emulator_bus::ReadWriteRegister<
+            u32,
+            registers_generated::flash_ctrl::bits::InterruptState::Register,
+        >,
+    ) {
+        // Interrupt state register: write 1 to clear
+        if _val.reg.is_set(InterruptState::Error) {
+            self.interrupt_state
+                .reg
+                .modify(InterruptState::Error::CLEAR);
+        }
+        if _val.reg.is_set(InterruptState::Event) {
+            self.interrupt_state
+                .reg
+                .modify(InterruptState::Event::CLEAR);
+        }
+    }
+
+    fn read_interrupt_enable(
+        &mut self,
+        _size: emulator_types::RvSize,
+    ) -> emulator_bus::ReadWriteRegister<
+        u32,
+        registers_generated::flash_ctrl::bits::InterruptEnable::Register,
+    > {
+        emulator_bus::ReadWriteRegister::new(self.interrupt_enable.reg.get())
+    }
+
+    fn write_interrupt_enable(
+        &mut self,
+        _size: emulator_types::RvSize,
+        _val: emulator_bus::ReadWriteRegister<
+            u32,
+            registers_generated::flash_ctrl::bits::InterruptEnable::Register,
+        >,
+    ) {
+        self.interrupt_enable.reg.set(_val.reg.get());
+    }
+
+    fn write_page_size(&mut self, _size: emulator_types::RvSize, _val: emulator_types::RvData) {
+        self.page_size.reg.set(_val);
+    }
+
+    // Return the page size of the flash storage connected to the controller
+    fn read_page_size(&mut self, _size: emulator_types::RvSize) -> emulator_types::RvData {
+        Self::PAGE_SIZE as u32
+    }
+
+    fn read_page_num(&mut self, _size: emulator_types::RvSize) -> emulator_types::RvData {
+        self.page_num.reg.get()
+    }
+
+    fn write_page_num(&mut self, _size: emulator_types::RvSize, _val: emulator_types::RvData) {
+        self.page_num.reg.set(_val);
+    }
+
+    fn read_page_addr(&mut self, _size: emulator_types::RvSize) -> emulator_types::RvData {
+        self.page_addr.reg.get()
+    }
+
+    fn write_page_addr(&mut self, _size: emulator_types::RvSize, _val: emulator_types::RvData) {
+        self.page_addr.reg.set(_val);
+    }
+
+    fn read_control(
+        &mut self,
+        _size: emulator_types::RvSize,
+    ) -> emulator_bus::ReadWriteRegister<
+        u32,
+        registers_generated::flash_ctrl::bits::Control::Register,
+    > {
+        emulator_bus::ReadWriteRegister::new(self.control.reg.get())
+    }
+
+    // Write to the control register to start the flash operation. The operation is processed in the next cycle.
+    fn write_control(
+        &mut self,
+        _size: emulator_types::RvSize,
+        _val: emulator_bus::ReadWriteRegister<
+            u32,
+            registers_generated::flash_ctrl::bits::Control::Register,
+        >,
+    ) {
+        self.control.reg.set(_val.reg.get());
+
+        // Set the operation_pending flag to true
+        self.operation_pending = true;
```

Comment:
You'll also want to set the `timer` here so that you are sure that the `poll()` method is called at the appropriate time.